source/bigint.c

```diff
@@ -0,0 +1,362 @@
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+#include <aws/common/bigint.h>
+
+#define BASE_BITS 32
+#define NIBBLES_PER_DIGIT ((BASE_BITS) / 4)
+#define LOWER_32_BIT_MASK 0xFFFFFFFF
+#define INT64_MIN_AS_HEX 0x8000000000000000
+
+/*
+ * Working set of invariants:
+ *
+ * (1) Negative zero is illegal
+ * (2) The only time leading (trailing) 0-value digits are allowed is a single instance to represent actual zero.
+ *
+ * Other proposals:
+ *
+ * (1) Functionally immutable API (no visible side-affects)
+ * (2) params may be temporarily altered during computation before being restored in order to prevent pointless copying
+ *   (most often around sign negation and operator switching)
+ *
+ * Dev ideas:
+ *
+ * (1) Internal arithmetic ops may use a "raw" interface to allow for subsequence operations used in multiply/divide.
+ */
+
+void aws_bigint_clean_up(struct aws_bigint *bigint) {
+    aws_array_list_clean_up(&bigint->digits);
+}
+
+static void s_advance_cursor_past_hex_prefix(struct aws_byte_cursor *hex_cursor) {
+    if (hex_cursor->len >= 2) {
+        const char *raw_ptr = (char *)hex_cursor->ptr;
+        if (raw_ptr[0] == '0' && (raw_ptr[1] == 'x' || raw_ptr[1] == 'X')) {
+            aws_byte_cursor_advance(hex_cursor, 2);
+        }
+    }
+}
+
+static void s_advance_cursor_to_non_zero(struct aws_byte_cursor *hex_cursor) {
+    while (hex_cursor->len > 0 && *hex_cursor->ptr == '0') {
+        aws_byte_cursor_advance(hex_cursor, 1);
+    }
+}
+
+static int s_uint32_from_hex(struct aws_byte_cursor digit_cursor, uint32_t *output_value) {
+    AWS_FATAL_ASSERT(digit_cursor.len <= NIBBLES_PER_DIGIT);
+
+    *output_value = 0;
+
+    while (digit_cursor.len > 0) {
+        char hex_digit = *digit_cursor.ptr;
+        uint32_t hex_value = 0;
+        if (hex_digit <= '9' && hex_digit >= '0') {
+            hex_value = hex_digit - '0';
+        } else if (hex_digit <= 'f' && hex_digit >= 'a') {
+            hex_value = hex_digit - 'a' + 10;
+        } else if (hex_digit <= 'F' && hex_digit >= 'A') {
+            hex_value = hex_digit - 'A' + 10;
+        } else {
+            return AWS_OP_ERR;
+        }
+
+        *output_value <<= 4;
+        *output_value += hex_value;
+
+        aws_byte_cursor_advance(&digit_cursor, 1);
+    }
+
+    return AWS_OP_SUCCESS;
+}
+
+int aws_bigint_init_from_hex(
+    struct aws_bigint *bigint,
+    struct aws_allocator *allocator,
+    struct aws_byte_cursor hex_digits) {
+
+    /* skip past the optional "0x" prefix */
+    s_advance_cursor_past_hex_prefix(&hex_digits);
+    if (hex_digits.len == 0) {
+        return aws_raise_error(AWS_ERROR_INVALID_ARGUMENT);
+    }
+
+    /* skip past leading zeros */
+    s_advance_cursor_to_non_zero(&hex_digits);
+    if (hex_digits.len == 0) {
+        return aws_bigint_init_from_uint64(bigint, allocator, 0);
+    }
+
+    uint64_t digit_count = hex_digits.len / BASE_BITS + 1;
```

Comment:
what aboutm 0x01ab?

wouldn't digit count end up being 2 instead of 3?

Reply:
Oops, should be "(hex_digits.len - 1) / NIBBLES_PER_DIGIT + 1" actually.  Every 8 hex digits is one of our digits, so 0-8 -> 1, 9-16->2, etc...